internal/linker/link_node_modules.ts

```diff
@@ -600,43 +531,143 @@ export type LinkerAliases = {
+        // module name to different targets in a non-sandboxed environment. The latter will lead to
+        // undeterministic behavior.
+        // TODO: can we detect the latter case and throw an apprioriate error?
+        log_verbose(`Out-of-date symlink for ${p} to ${symlinkPath} detected. Target should be ${
+            target}. Unlinking.`);
+        await unlink(p);
+      }
+    }
+    return symlink(target, p);
+  }
+
+  // Symlink all node_modules roots defined. These are 3rd party deps in external npm workspaces
+  // lined to node_modules folders at the root or in sub-directories
+  for (const packagePath of Object.keys(roots)) {
+    const workspace = roots[packagePath];
+    const workspacePath =
+        await resolveExternalWorkspacePath(workspace, startCwd, isExecroot, execroot, runfiles);
+    log_verbose(`resolved ${workspace} workspace path to ${workspacePath}`);
+    const workspaceNodeModules = `${workspacePath}/node_modules`;
+    if (packagePath) {
+      // sub-directory node_modules
+      if (await exists(workspaceNodeModules)) {
+        let resolvedPackagePath;
+        if (await exists(packagePath)) {
+          await symlinkWithUnlink(workspaceNodeModules, `${packagePath}/node_modules`);
+          resolvedPackagePath = packagePath;
+        }
+        if (!isExecroot) {
+          // Under runfiles, we symlink into the package in runfiles as well.
+          // When inside the sandbox, the execroot location will not exist to symlink to.
+          const runfilesPackagePath = `${startCwd}/${packagePath}`;
+          if (await exists(runfilesPackagePath)) {
+            if (resolvedPackagePath) {
+              await symlinkWithUnlink(
+                  `${resolvedPackagePath}/node_modules`, `${runfilesPackagePath}/node_modules`);
+            } else {
+              await symlinkWithUnlink(workspaceNodeModules, `${runfilesPackagePath}/node_modules`);
+            }
+            resolvedPackagePath = runfilesPackagePath;
+          }
+        }
+        const packagePathBin = `${bin}/${packagePath}`;
+        if (resolvedPackagePath && await exists(packagePathBin)) {
+          // if bin path exists, symlink bin/package/node_modules -> package/node_modules
+          await symlinkWithUnlink(
+              `${resolvedPackagePath}/node_modules`, `${packagePathBin}/node_modules`);
+        }
+      }
+    } else {
+      // root node_modules
+      if (await exists(workspaceNodeModules)) {
+        await symlinkWithUnlink(workspaceNodeModules, `node_modules`);
+      } else {
+        // Special case if there no target to symlink to for the root workspace, create a
+        // root node_modules folder for 1st party deps
+        log_verbose(
+            'no root npm workspace node_modules folder to link to; creating node_modules directory in',
+            process.cwd());
+        await mkdirp('node_modules');
+      }
+    }
+  }
+
+  if (!roots || !roots['']) {
```

Comment:
wondering why roots would be truthy and then we want to check if it has an empty string element. i guess there's some logic somewhere that treats no third-party modules as empty? feels like it should be fixed there (don't put an empty string in the roots)